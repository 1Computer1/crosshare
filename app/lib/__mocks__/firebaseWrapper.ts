const firebaseTesting = jest.requireActual('@firebase/testing');
import type firebaseAdminType from 'firebase-admin';

export let App: firebase.app.App;
export let AdminApp: firebaseAdminType.app.App;

export const setApp = (app: firebase.app.App) => { App = app; };
export const setAdminApp = (app: firebaseAdminType.app.App) => { AdminApp = app; };

export const AuthProvider = null;

export const DeleteSentinal = firebaseTesting.firestore.FieldValue.delete();

export const ServerTimestamp = firebaseTesting.firestore.FieldValue.serverTimestamp();

export const TimestampClass = firebaseTesting.firestore.Timestamp;
export const AdminTimestamp = firebaseTesting.firestore.Timestamp;

let anonApp: firebase.app.App;
let anonUser: firebase.User;
export const setUpForSignInAnonymously = (app: firebase.app.App, user: firebase.User) => {
  anonApp = app;
  anonUser = user;
};

export const signInAnonymously = async () => {
  setApp(anonApp);
  return anonUser;
};
