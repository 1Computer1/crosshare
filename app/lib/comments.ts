import {
  CommentForModerationWithIdT,
  CommentWithRepliesT,
  DBPuzzleT,
  DBPuzzleV,
} from './dbtypes';
import { getFromDB } from './dbUtils';

function findCommentById(
  comments: Array<CommentWithRepliesT>,
  id: string
): CommentWithRepliesT | null {
  for (const comment of comments) {
    if (comment.i === id) {
      return comment;
    }
    if (comment.r !== undefined) {
      const res = findCommentById(comment.r, id);
      if (res !== null) {
        return res;
      }
    }
  }
  return null;
}

export async function moderateComments(
  db: FirebaseFirestore.Firestore,
  commentsForModeration: Array<CommentForModerationWithIdT>,
  commentIdsForDeletion: Set<string>,
  autoModerating: boolean
) {
  const puzzles: Record<string, DBPuzzleT> = {};
  if (commentsForModeration) {
    for (const comment of commentsForModeration) {
      // Don't need to do anything for any comment that has been marked for deletion
      if (!commentIdsForDeletion.has(comment.i)) {
        let puzzle: DBPuzzleT | null = null;
        const fromCache = puzzles[comment.pid];
        if (fromCache) {
          puzzle = fromCache;
        } else {
          try {
            puzzle = await getFromDB('c', comment.pid, DBPuzzleV);
            puzzles[comment.pid] = puzzle;
          } catch {
            puzzle = null;
          }
        }
        if (puzzle) {
          if (puzzle.cs === undefined) {
            puzzle.cs = [];
          }
          if (comment.rt === null) {
            puzzle.cs.push(comment);
          } else {
            const parent = findCommentById(puzzle.cs, comment.rt);
            if (parent === null) {
              throw new Error('parent comment not found');
            }
            if (parent.r) {
              parent.r.push(comment);
            } else {
              parent.r = [comment];
            }
          }
        }
      }
      await db.collection('cfm').doc(comment.i).delete();
      if (autoModerating) {
        db.collection('automoderated').doc(comment.i).create(comment);
      }
    }
  }

  // Now we've merged in all the comments, so update the puzzles:
  for (const [puzzleId, dbPuzzle] of Object.entries(puzzles)) {
    console.log('updating', puzzleId, dbPuzzle.cs);
    await db.collection('c').doc(puzzleId).update({ cs: dbPuzzle.cs });
  }
}
