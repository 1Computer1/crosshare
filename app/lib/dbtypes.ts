import * as t from 'io-ts';

// We import this instead of defining it here so that we can define it
// differently for the main react app and the 'functions' subdirectory
import { timestamp } from './timestamp';
export { timestamp } from './timestamp';

const CommentV = t.type({
  /** comment text */
  c: t.string,
  /** author id */
  a: t.string,
  /** author display name */
  n: t.string,
  /** author solve time in fractional seconds */
  t: t.number,
  /** author did cheat? */
  ch: t.boolean,
  /** comment publish timestamp */
  p: timestamp
});
type CommentT = t.TypeOf<typeof CommentV>;

export interface CommentWithRepliesT extends CommentT {
  /** comment id */
  i: string,
  /** replies */
  r?: Array<CommentWithRepliesT>
}

const CommentWithRepliesV: t.Type<CommentWithRepliesT> = t.recursion('CommentWithReplies', () =>
  t.intersection([
    CommentV,
    t.type({
      /** comment id */
      i: t.string,
    }),
    t.partial({
      /** replies */
      r: t.array(CommentWithRepliesV)
    })
  ])
);

export const CommentForModerationV = t.intersection([
  CommentV,
  t.type({
    /** puzzle id */
    pid: t.string,
    /** id of the comment this is a reply to */
    rt: t.union([t.string, t.null])
  })
]);
export type CommentForModerationT = t.TypeOf<typeof CommentForModerationV>;

export const CommentForModerationWithIdV = t.intersection([
  CommentForModerationV,
  t.type({
    /** comment id */
    i: t.string,
  })
]);
export type CommentForModerationWithIdT = t.TypeOf<typeof CommentForModerationWithIdV>;

const DBPuzzleMandatoryV = t.type({
  /** created at */
  ca: timestamp,
  /** author's user id */
  a: t.string,
  /** author's display name */
  n: t.string,
  /** category */
  c: t.union([t.string, t.null]),
  /** is this puzzle moderated? */
  m: t.boolean,
  /** timestamp when the puzzle goes live */
  p: t.union([timestamp, t.null]),
  /** title */
  t: t.string,
  /** grid width / columns */
  w: t.number,
  /** grid height / rows */
  h: t.number,
  /** across clue strings */
  ac: t.array(t.string),
  /** across clue display numbers */
  an: t.array(t.number),
  /** down clue strings */
  dc: t.array(t.string),
  /** down clue display numbers */
  dn: t.array(t.number),
  /** grid (solution) */
  g: t.array(t.string),
});
const DBPuzzleOptionalV = t.partial({
  /** highlighted cell indexes */
  hs: t.array(t.number),
  /** use shade instead of circle for highlight? */
  s: t.boolean,
  /** comments */
  cs: t.array(CommentWithRepliesV),
  /** is this puzzle featured (should appear on homepage) */
  f: t.boolean,
});
export const DBPuzzleV = t.intersection([DBPuzzleMandatoryV, DBPuzzleOptionalV]);
export type DBPuzzleT = t.TypeOf<typeof DBPuzzleV>;

export const TimestampedPuzzleV = downloadTimestamped(DBPuzzleV);
export type TimestampedPuzzleT = t.TypeOf<typeof TimestampedPuzzleV>;

const PlayBaseV = t.type({
  /** crossword id */
  c: t.string,
  /** updated at */
  ua: timestamp,
  /** filled in grid */
  g: t.array(t.string),
  /** play time of last update to each cell, in fractional seconds */
  ct: t.array(t.number),
  /** update iteration count per cell */
  uc: t.array(t.number),
  /** list of (checked) correct cells */
  vc: t.array(t.number),
  /** list of (checked) wrong cells */
  wc: t.array(t.number),
  /** list of cells ever marked wrong */
  we: t.array(t.number),
  /** list of revealed cells */
  rc: t.array(t.number),
  /** total play time, in fractional seconds */
  t: t.number,
  /** did "cheat"? */
  ch: t.boolean,
  /** finished the puzzle? */
  f: t.boolean,
});

export const LegacyPlayV = t.intersection([
  PlayBaseV,
  t.type({
    /** user id */
    u: t.string,
  }),
  t.partial({
    /** puzzle title, optional for legacy plays in the db */
    n: t.string,
  })
]);
export type LegacyPlayT = t.TypeOf<typeof LegacyPlayV>;

export const PlayV = t.intersection([
  PlayBaseV,
  t.type({
    /** user id */
    u: t.string,
    /** puzzle title */
    n: t.string,
  })
]);
export type PlayT = t.TypeOf<typeof PlayV>;

// We don't need a user id in local storage, but we do need a title
export const PlayWithoutUserV = t.intersection([
  PlayBaseV,
  t.type({
    /** puzzle title */
    n: t.string,
  })
]);
export type PlayWithoutUserT = t.TypeOf<typeof PlayWithoutUserV>;

export function downloadTimestamped<A>(type: t.Type<A>) {
  return t.type({
    downloadedAt: timestamp,
    data: type,
  });
}

export function downloadOptionallyTimestamped<A>(type: t.Type<A>) {
  return t.type({
    downloadedAt: t.union([timestamp, t.null]),
    data: type,
  });
}

export function getDateString(pd: Date) {
  return pd.getUTCFullYear() + '-' + pd.getUTCMonth() + '-' + pd.getUTCDate();
}

export function prettifyDateString(dateString: string) {
  const groups = dateString.match(/^(\d+)-(\d+)-(\d+)$/);
  if (!groups) {
    throw new Error('Bad date string: ' + dateString);
  }
  return (parseInt(groups[2]) + 1) + '/' + parseInt(groups[3]) + '/' + groups[1];
}

export const PuzzleStatsV = t.type({
  /** author id, denormalized for security rules purposes. */
  a: t.string,
  /** updated at */
  ua: timestamp,
  /** total completions */
  n: t.number,
  /** total completions without cheats */
  s: t.number,
  /** total time spent on puzzle, in fractional seconds */
  nt: t.number,
  /** total time spent by those w/o cheats */
  st: t.number,
  /** total play time of last update to each cell, in fractional seconds */
  ct: t.array(t.number),
  /** total update iteration count for each cell */
  uc: t.array(t.number),
});
export type PuzzleStatsT = t.TypeOf<typeof PuzzleStatsV>;

export const DailyStatsV = t.type({
  /** updated at */
  ua: timestamp,
  /** total completions */
  n: t.number,
  /** user ids with completions */
  u: t.array(t.string),
  /** completions by puzzleId */
  c: t.record(t.string, t.number),
  /** completions by hour (as UTC 0-23) */
  h: t.array(t.number),
});
export type DailyStatsT = t.TypeOf<typeof DailyStatsV>;

export const CronStatusV = t.type({
  ranAt: timestamp,
});
export type CronStatusT = t.TypeOf<typeof CronStatusV>;

/** created at, title */
const AuthoredPuzzleV = t.tuple([timestamp, t.string]);
export type AuthoredPuzzleT = t.TypeOf<typeof AuthoredPuzzleV>;

/** keys are puzzle ids */
export const AuthoredPuzzlesV = t.record(t.string, AuthoredPuzzleV);

/** date string -> puzzle id */
export const CategoryIndexV = t.record(t.string, t.string);
export type CategoryIndexT = t.TypeOf<typeof CategoryIndexV>;
