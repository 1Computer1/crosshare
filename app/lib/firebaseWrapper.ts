declare var firebase: typeof import('firebase');

// On the server we need to require firebase - on client it's done w/ script tags
if (typeof window === "undefined") {
  globalThis.firebase = require("firebase");
  require("firebase/firestore");
}

const firebaseConfig = {
  apiKey: "AIzaSyBrmmBf91peVT5T_Z7N3z9oizsPH5u2pUc",
  authDomain: "auth.crosshare.org",
  databaseURL: "https://mdcrosshare.firebaseio.com",
  projectId: "mdcrosshare",
  storageBucket: "mdcrosshare.appspot.com",
  messagingSenderId: "603173482014",
  appId: "1:603173482014:web:98d7d820731b7c5eaa080f",
  measurementId: "G-LTLN7Z4XBS"
};

// Initialize Firebase
export let App: firebase.app.App;
if (firebase.apps.length) {
  App = firebase.apps[0];
} else {
  App = firebase.initializeApp(firebaseConfig);
  if (typeof window !== "undefined") {
    if (process.env.NODE_ENV === 'production') {
      App.performance();
    }
  }
  if (process.env.NEXT_PUBLIC_USE_FIREBASE_EMULATOR) {
    var db = App.firestore();
    db.settings({
      host: "localhost:8080",
      ssl: false
    });
  }
}

export const setApp = (app: firebase.app.App) => { App = app };

export const AuthProvider = new firebase.auth.GoogleAuthProvider();

export const DeleteSentinal = firebase.firestore.FieldValue.delete();

export const TimestampClass = firebase.firestore.Timestamp;
export type TimestampType = firebase.firestore.Timestamp;
