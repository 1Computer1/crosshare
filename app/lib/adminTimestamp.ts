/* this is the firebase timestamp type for the main react project! */

import * as t from 'io-ts';
import { either } from 'fp-ts/lib/Either';

import { AdminTimestamp } from './firebaseWrapper';
import type firebaseAdminType from 'firebase-admin';

const isFirestoreTimestamp = (u: unknown): u is firebaseAdminType.firestore.Timestamp =>
  u ? u instanceof AdminTimestamp : false;

const validateTimestamp: t.Validate<unknown, firebaseAdminType.firestore.Timestamp> = (i, c) => {
  if (isFirestoreTimestamp(i)) {
    return t.success(i);
  }
  return either.chain(
    t.type({ seconds: t.number, nanoseconds: t.number }).validate(i, c),
    obj => t.success(new AdminTimestamp(obj.seconds, obj.nanoseconds))
  );
};

export const adminTimestamp = new t.Type<firebaseAdminType.firestore.Timestamp>(
  'AdminTimestamp',
  isFirestoreTimestamp,
  validateTimestamp,
  t.identity,
);
