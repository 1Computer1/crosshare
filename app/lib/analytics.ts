/* This gets used by the analytics cron function in the `functions` directory.

It lives here so we can test it. */
import type firebase from 'firebase-admin';
import { PathReporter } from 'io-ts/lib/PathReporter';
import { isRight } from 'fp-ts/lib/Either';

import {
  LegacyPlayV,
  PuzzleStatsT,
  PuzzleStatsV,
  DailyStatsV,
  DailyStatsT,
  DBPuzzleV,
  DBPuzzleT,
  getDateString,
} from './dbtypes';

import { TimestampType } from './firebaseWrapper';

export async function runAnalytics(
  db: firebase.firestore.Firestore,
  startTimestamp: TimestampType,
  endTimestamp: TimestampType
) {
  console.log(
    'Updating analytics btwn',
    startTimestamp.toDate().toLocaleString(),
    endTimestamp.toDate().toLocaleString()
  );
  const value = await db
    .collection('p')
    .where('f', '==', true)
    .where('ua', '>=', startTimestamp)
    .where('ua', '<', endTimestamp)
    .orderBy('ua', 'asc')
    .get();
  console.log('Updating analytics for ' + value.size + ' plays');

  const puzzleMap: Map<string, DBPuzzleT> = new Map();
  const puzzleStatsMap: Map<string, PuzzleStatsT> = new Map();
  const dailyStatsMap: Map<string, DailyStatsT> = new Map();

  // Get puzzle obj from cache or db
  async function getPuzzle(puzzleId: string): Promise<DBPuzzleT> {
    const puzzle = puzzleMap.get(puzzleId);
    if (puzzle) {
      return puzzle;
    }
    const puzzleRes = await db.collection('c').doc(puzzleId).get();
    if (!puzzleRes.exists) {
      throw new Error('Missing puzzle but have play: ' + puzzleId);
    }
    const dbpuzzle = DBPuzzleV.decode(puzzleRes.data());
    if (!isRight(dbpuzzle)) {
      console.error(PathReporter.report(dbpuzzle).join(','));
      throw new Error('Malformed puzzle');
    }
    puzzleMap.set(puzzleId, dbpuzzle.right);
    return dbpuzzle.right;
  }

  for (const doc of value.docs) {
    const validationResult = LegacyPlayV.decode(doc.data());
    if (!isRight(validationResult)) {
      console.error(PathReporter.report(validationResult).join(','));
      throw new Error('Malformed play');
    }
    const play = validationResult.right;

    // Make sure we have a puzzle stats obj in the cache for the relevant puzzle
    let puzzleStats = puzzleStatsMap.get(play.c);
    if (!puzzleStats) {
      // get puzzle stats from db or create a new stats object
      const psvalue = await db.collection('s').doc(play.c).get();
      if (psvalue.exists) {
        const result = PuzzleStatsV.decode(psvalue.data());
        if (!isRight(result)) {
          console.error(PathReporter.report(result).join(','));
          throw new Error('Malformed puzzle stats');
        }
        puzzleStats = result.right;
        puzzleStatsMap.set(play.c, puzzleStats);
      } else {
        const puzzle = await getPuzzle(play.c);
        puzzleStats = {
          a: puzzle.a,
          ua: endTimestamp,
          n: 0,
          s: 0,
          nt: 0,
          st: 0,
          ct: [],
          uc: [],
          sct: Array.from(
            { length: 6 },
            () => Math.random().toString(36)[2]
          ).join(''),
        };
        puzzleStatsMap.set(play.c, puzzleStats);
      }
    }

    // don't count the author's play in puzzle stats
    if (puzzleStats.a === play.u) {
      continue;
    }

    puzzleStats.ua = endTimestamp;
    puzzleStats.n += 1;
    puzzleStats.nt += play.t;
    if (!play.ch) {
      puzzleStats.s += 1;
      puzzleStats.st += play.t;
    }
    const maxTime = Math.max(...play.ct);
    const maxCount = Math.max(...play.uc);
    for (let i = 0; i < play.ct.length; i += 1) {
      let updateTime = play.ct[i];
      const updateIters = play.uc[i];
      if (play.rc.indexOf(i) !== -1 || play.we.indexOf(i) !== -1) {
        /* If a cell was revealed or checked & wrong, make it's update time the
         * end of the play. This way cheat cells always show as taking the
         * longest for the user in question. */
        updateTime = play.t;
      }
      puzzleStats.ct[i] = (puzzleStats.ct[i] || 0) + updateTime / maxTime;
      puzzleStats.uc[i] = (puzzleStats.uc[i] || 0) + updateIters / maxCount;
    }

    // Next update daily stats for the relevant date
    const pd = play.ua.toDate();
    const dateString = getDateString(pd);
    let dailyStats = dailyStatsMap.get(dateString);
    if (!dailyStats) {
      // get daily stats from db or create a new stats object
      const dsvalue = await db.collection('ds').doc(dateString).get();
      if (dsvalue.exists) {
        const result = DailyStatsV.decode(dsvalue.data());
        if (!isRight(result)) {
          console.error(PathReporter.report(result).join(','));
          throw new Error('Malformed daily stats');
        }
        dailyStats = result.right;
        dailyStatsMap.set(dateString, dailyStats);
      } else {
        dailyStats = {
          ua: endTimestamp,
          n: 0,
          u: [],
          c: {},
          i: {},
          h: Array<number>(24).fill(0),
        };
        dailyStatsMap.set(dateString, dailyStats);
      }
    }
    dailyStats.ua = endTimestamp;
    dailyStats.n += 1;
    if (dailyStats.u.indexOf(play.u) === -1) {
      dailyStats.u.push(play.u);
    }
    dailyStats.c[play.c] = (dailyStats.c[play.c] || 0) + 1;
    if (dailyStats.i && !dailyStats.i[play.c]) {
      const puzzle = await getPuzzle(play.c);
      dailyStats.i[play.c] = [puzzle.t, puzzle.n, puzzle.a];
    }
    const hour = pd.getUTCHours();
    dailyStats.h[hour] = (dailyStats.h[hour] || 0) + 1;
  }

  console.log('Done, writing ' + puzzleStatsMap.size + ' puzzle stats objects');
  for (const [crosswordId, puzzleStats] of puzzleStatsMap.entries()) {
    await db.collection('s').doc(crosswordId).set(puzzleStats);
  }
  console.log('Writing ' + dailyStatsMap.size + ' daily stats objects');
  for (const [dateString, dailyStats] of dailyStatsMap.entries()) {
    await db.collection('ds').doc(dateString).set(dailyStats);
  }
}
