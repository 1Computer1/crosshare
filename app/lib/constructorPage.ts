import * as t from 'io-ts';
import { isRight } from 'fp-ts/lib/Either';
import { PathReporter } from 'io-ts/lib/PathReporter';

import { timestamp } from './timestamp';

export const CONSTRUCTOR_PAGE_COLLECTION = 'cp';

const ConstructorPageV = t.type({
  /** user id */
  u: t.string,
  /** display name */
  n: t.string,
  /** author bio */
  b: t.string,
  /** timestamp when last updated */
  t: timestamp,
});
export interface ConstructorPageT extends Omit<t.TypeOf<typeof ConstructorPageV>, 't'> {
  id: string,
}

export function validate(cp: unknown, username: string): ConstructorPageT | null {
  const validationResult = ConstructorPageV.decode(cp);
  if (isRight(validationResult)) {
    const { t, ...partial } = validationResult.right;  // eslint-disable-line @typescript-eslint/no-unused-vars
    return { ...partial, id: username };
  } else {
    console.error(PathReporter.report(validationResult).join(','));
    return null;
  }
}
