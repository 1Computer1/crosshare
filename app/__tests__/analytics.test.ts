import { TimestampClass } from '../lib/firebaseWrapper';
import * as firebaseTesting from '@firebase/testing';
import { PlayT, LegacyPlayT, DBPuzzleT } from '../lib/dbtypes';
import { runAnalytics } from '../lib/analytics';

let adminApp: firebase.app.App;

let play1: PlayT;
let play2: LegacyPlayT;
let play3: LegacyPlayT;

jest.mock('../lib/firebaseWrapper');

const projectId = 'analyticstest';

beforeAll(async () => {
  adminApp = firebaseTesting.initializeAdminApp({ projectId }) as firebase.app.App;
});

beforeEach(async () => {
  await firebaseTesting.clearFirestoreData({ projectId });

  const thirtyAgo = new Date();
  thirtyAgo.setMinutes(thirtyAgo.getMinutes() - 30);
  const twentyAgo = new Date();
  twentyAgo.setMinutes(twentyAgo.getMinutes() - 20);

  play1 = {
    c: 'mike',
    u: 'blah',
    ua: TimestampClass.fromDate(twentyAgo),
    g: [],
    ct: [1, 2, 4, 100, 100].concat(new Array(20).fill(100)),
    uc: [5, 2].concat(new Array(23).fill(1)),
    vc: [],
    wc: [],
    we: [],
    rc: [],
    t: 70,
    ch: true,
    f: true,
    n: 'Puzzle title'
  };
  await adminApp.firestore().collection('p').doc('mike-blah').set(play1);

  play2 = {
    c: 'mike',
    u: 'anonymous-user-id',
    ua: TimestampClass.fromDate(thirtyAgo),
    g: [],
    ct: [],
    uc: [],
    vc: [],
    wc: [],
    we: [],
    rc: [],
    t: 44,
    ch: false,
    f: true,
  };
  await adminApp.firestore().collection('p').doc('mike-anonymous-user-id').set(play2);
  // Since play2 is a LegacyPlayT, getPlays() will look up the puzzle to get a title
  const puzzle: DBPuzzleT = {
    c: 'dailymini',
    m: true,
    t: 'Raises, as young',
    dn: [1, 2, 3, 4, 5],
    ac:
      [' Cobbler\'s forms',
        'Absolutely perfect',
        'Spike Lee\'s "She\'s ___ Have It"',
        'English class assignment',
        'Raises, as young'],
    dc:
      ['Hybrid whose father is a lion',
        '___ of reality (wake-up call)',
        '___ date (makes wedding plans)',
        'Middle Ages invader',
        'Has a great night at the comedy club'],
    p: TimestampClass.fromDate(new Date('6/10/2020')),
    a: 'fSEwJorvqOMK5UhNMHa4mu48izl1',
    an: [1, 6, 7, 8, 9],
    g:
      ['L',
        'A',
        'S',
        'T',
        'S',
        'I',
        'D',
        'E',
        'A',
        'L',
        'G',
        'O',
        'T',
        'T',
        'A',
        'E',
        'S',
        'S',
        'A',
        'Y',
        'R',
        'E',
        'A',
        'R',
        'S'],
    h: 5,
    w: 5,
    cs:
      [{
        c:
          'A couple of two-worders today which I don\'t love, but I hope you all got it anyway!',
        i: 'LwgoVx0BAskM4wVJyoLj',
        t: 36.009,
        p: TimestampClass.now(),
        a: 'fSEwJorvqOMK5UhNMHa4mu48izl1',
        n: 'Mike D',
        ch: false,
      }],
    n: 'Mike D'
  };
  await adminApp.firestore().collection('c').doc('mike').set(puzzle);

  play3 = {
    c: 'mike',
    u: 'other-user-id',
    ua: TimestampClass.fromDate(twentyAgo),
    g: [],
    ct: [],
    uc: [],
    vc: [],
    wc: [],
    we: [],
    rc: [],
    t: 44,
    ch: false,
    f: true
  };
  await adminApp.firestore().collection('p').doc('mike-other-user-id').set(play3);
});

afterAll(async () => {
  await adminApp.delete();
});

test('run for all time w/o initial state', async () => {
  const hourAgo = new Date();
  hourAgo.setMinutes(hourAgo.getMinutes() - 60);

  await runAnalytics(adminApp.firestore(), TimestampClass.fromDate(hourAgo), TimestampClass.fromDate(new Date()));

  const res = await adminApp.firestore().collection('ds').get();
  expect(res.size).toEqual(1);
  // Can't snapshot updatedAt or playcount by hour
  const { ua, h, ...toSnapshot } = res.docs[0].data();
  expect(ua).not.toBeFalsy();
  expect(h.length).toEqual(24);
  expect(toSnapshot).toMatchSnapshot();

  const pres = await adminApp.firestore().collection('s').doc('mike').get();
  const data = pres.data();
  if (data === undefined) {
    throw new Error('botch');
  }
  const { ua: pua, ...pToSnapshot } = data;
  expect(pua).not.toBeFalsy();
  expect(pToSnapshot).toMatchSnapshot();
});

test('run for more recent w/o initial state', async () => {
  const twentyFive = new Date();
  twentyFive.setMinutes(twentyFive.getMinutes() - 25);

  await runAnalytics(adminApp.firestore(), TimestampClass.fromDate(twentyFive), TimestampClass.fromDate(new Date()));

  const res = await adminApp.firestore().collection('ds').get();
  expect(res.size).toEqual(1);
  // Can't snapshot updatedAt or playcount by hour
  const { ua, h, ...toSnapshot } = res.docs[0].data();
  expect(ua).not.toBeFalsy();
  expect(h.length).toEqual(24);
  expect(toSnapshot).toMatchSnapshot();

  const pres = await adminApp.firestore().collection('s').doc('mike').get();
  const data = pres.data();
  if (data === undefined) {
    throw new Error('botch');
  }
  const { ua: pua, ...pToSnapshot } = data;
  expect(pua).not.toBeFalsy();
  expect(pToSnapshot).toMatchSnapshot();
});
