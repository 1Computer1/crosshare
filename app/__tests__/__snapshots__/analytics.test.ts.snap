// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run for all time w/ some meta submissions 1`] = `
Object {
  "c": Object {
    "mike": 3,
  },
  "i": Object {
    "mike": Array [
      "Raises, as young",
      "Mike D",
      "fSEwJorvqOMK5UhNMHa4mu48izl1",
    ],
  },
  "n": 3,
  "u": Array [
    "anonymous-user-id",
    "blah",
    "other-user-id",
  ],
}
`;

exports[`run for all time w/ some meta submissions 2`] = `
Object {
  "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  "ct": Array [
    0.01,
    0.02,
    0.04,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
  ],
  "ct_subs": Array [
    Object {
      "e": "foo@example.com",
      "n": "Blah",
      "s": "just a guess",
      "t": null,
      "u": "blah",
    },
    Object {
      "e": null,
      "n": "Mike D",
      "s": "my submission",
      "t": null,
      "u": "other-user-id",
    },
  ],
  "n": 3,
  "nt": 158,
  "s": 2,
  "st": 88,
  "uc": Array [
    1,
    0.4,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
  ],
}
`;

exports[`run for all time w/ some meta submissions 3`] = `
Array [
  Object {
    "n": "Blah",
    "s": "just a guess",
    "t": null,
  },
  Object {
    "n": "Mike D",
    "s": "my submission",
    "t": null,
  },
]
`;

exports[`run for all time w/ some meta submissions 4`] = `
Object {
  "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  "ct": Array [
    0.02,
    0.04,
    0.08,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
  ],
  "ct_subs": Array [
    Object {
      "e": "foo@example.com",
      "n": "Blah",
      "s": "just a guess",
      "t": null,
      "u": "blah",
    },
    Object {
      "e": null,
      "n": "Mike D",
      "s": "my submission",
      "t": null,
      "u": "other-user-id",
    },
    Object {
      "e": "foo@example.com",
      "n": "Blah",
      "s": "just a guess 2",
      "t": null,
      "u": "blah",
    },
    Object {
      "e": null,
      "n": "Mike D",
      "s": "my submission 2",
      "t": null,
      "u": "other-user-id",
    },
  ],
  "n": 6,
  "nt": 316,
  "s": 4,
  "st": 176,
  "uc": Array [
    2,
    0.8,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
  ],
}
`;

exports[`run for all time w/ some meta submissions 5`] = `
Array [
  Object {
    "n": "Blah",
    "s": "just a guess",
    "t": null,
  },
  Object {
    "n": "Mike D",
    "s": "my submission",
    "t": null,
  },
  Object {
    "n": "Blah",
    "s": "just a guess 2",
    "t": null,
  },
  Object {
    "n": "Mike D",
    "s": "my submission 2",
    "t": null,
  },
]
`;

exports[`run for all time w/o initial state 1`] = `
Object {
  "c": Object {
    "mike": 3,
  },
  "i": Object {
    "mike": Array [
      "Raises, as young",
      "Mike D",
      "fSEwJorvqOMK5UhNMHa4mu48izl1",
    ],
  },
  "n": 3,
  "u": Array [
    "anonymous-user-id",
    "blah",
    "other-user-id",
  ],
}
`;

exports[`run for all time w/o initial state 2`] = `
Object {
  "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  "ct": Array [
    0.01,
    0.02,
    0.04,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
  ],
  "n": 3,
  "nt": 158,
  "s": 2,
  "st": 88,
  "uc": Array [
    1,
    0.4,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
  ],
}
`;

exports[`run for more recent w/o initial state 1`] = `
Object {
  "c": Object {
    "mike": 2,
  },
  "i": Object {
    "mike": Array [
      "Raises, as young",
      "Mike D",
      "fSEwJorvqOMK5UhNMHa4mu48izl1",
    ],
  },
  "n": 2,
  "u": Array [
    "blah",
    "other-user-id",
  ],
}
`;

exports[`run for more recent w/o initial state 2`] = `
Object {
  "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  "ct": Array [
    0.01,
    0.02,
    0.04,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
  ],
  "n": 2,
  "nt": 114,
  "s": 1,
  "st": 44,
  "uc": Array [
    1,
    0.4,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
    0.2,
  ],
}
`;

exports[`run w/ initial state 1`] = `
Object {
  "c": Object {
    "mike": 6,
  },
  "i": Object {
    "mike": Array [
      "Raises, as young",
      "Mike D",
      "fSEwJorvqOMK5UhNMHa4mu48izl1",
    ],
  },
  "n": 6,
  "u": Array [
    "anonymous-user-id",
    "blah",
    "other-user-id",
  ],
}
`;

exports[`run w/ initial state 2`] = `
Object {
  "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  "ct": Array [
    0.02,
    0.04,
    0.08,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
  ],
  "n": 6,
  "nt": 316,
  "s": 4,
  "st": 176,
  "uc": Array [
    2,
    0.8,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
    0.4,
  ],
}
`;
